src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -1207,28 +1213,126 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
+    } else {
+      version = source->get_object_version();
+    }
+    
+    prefix = source->get_bucket()->get_name() + "_" + version + "_" + obj.objName;
+    policy_prefix = prefix;
+
+    auto blockDir = source->driver->get_block_dir(); 
+    rgw::d4n::CacheBlock block;
+    block.cacheObj.objName = source->get_name();
+    block.cacheObj.bucketName = source->get_bucket()->get_name();
+    block.blockID = 0;
+    block.size = 0;
+
+    // retrieve head object from directory to check for dirty flag
+    if ((ret = blockDir->get(dpp, &block, y)) < 0) { // Sam: can we somehow return the block from the check_head_exists_in_cache_get_oid method to avoid this redis call?
+      ldpp_dout(dpp, 0) << "Failed to retrieve head object in block directory for: " << obj.objName << ", ret=" << ret << dendl;
+      return next->delete_obj(dpp, y, flags);
+    }
+
+    if (block.dirty == true)
+      prefix = "D_" + prefix;
+
+    if ((ret = source->driver->get_cache_driver()->delete_data(dpp, prefix, y)) == 0) { // Sam: do we want del or delete_data here? 
+      if ((ret = source->driver->get_policy_driver()->get_cache_policy()->erase(dpp, policy_prefix, y))) {
+	if ((ret = blockDir->del(dpp, &block, y)) < 0) {
+	  ldpp_dout(dpp, 0) << "Failed to delete head object in block directory for: " << obj.objName << ", ret=" << ret << dendl;
+	  return next->delete_obj(dpp, y, flags);
+        }
+      } else {
+	ldpp_dout(dpp, 0) << "Failed to delete head policy entry for: " << obj.objName << ", ret=" << ret << dendl;
+	return next->delete_obj(dpp, y, flags);
+      }
+    } else {
+      ldpp_dout(dpp, 0) << "Failed to delete head object for: " << obj.objName << ", ret=" << ret << dendl;
+      return next->delete_obj(dpp, y, flags);
+    }
 
-  if ((ret = source->driver->get_cache_driver()->del(dpp, source->get_key().get_oid(), y)) < 0) 
-    ldpp_dout(dpp, 0) << "D4NFilterObject::" << __func__ << "(): CacheDriver del() method failed, ret=" << ret << dendl;
+    if ((ret = source->driver->get_obj_dir()->del(dpp, &obj, y)) < 0) {
+      ldpp_dout(dpp, 0) << "Failed to delete object directory entry for: " << obj.objName << ", ret=" << ret << dendl;
+      return next->delete_obj(dpp, y, flags);
+    }
+    
+    block.cacheObj.objName = source->get_key().get_oid();
+    off_t lst = source->get_obj_size();
+    off_t fst = 0;
+
+    do {
+      prefix = source->get_bucket()->get_name() + "_" + version + "_" + obj.objName;
+      if (fst >= lst) {
+	break;
+      }
+
+      off_t cur_size = std::min<off_t>(fst + dpp->get_cct()->_conf->rgw_max_chunk_size, lst);
+      off_t cur_len = cur_size - fst;
+      block.blockID = static_cast<uint64_t>(fst);
+      block.size = static_cast<uint64_t>(cur_len);
+
+      if ((ret = blockDir->get(dpp, &block, y)) < 0) {
+	ldpp_dout(dpp, 0) << "Failed to retrieve directory entry for: " << obj.objName << " blockid: " << fst << " block size: " << cur_len << ", ret=" << ret << dendl;
+	return next->delete_obj(dpp, y, flags);
```

Comment:
at this point the loop must continue for other blocks in case they are in the cache.

Reply:
same question as below in regards to a block deletion failing